### PR TITLE
feat: connection lookup logic moved out of didexhange service

### LINKED ISSUE #1015: Move ConnectionRecorder to common service package

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- moved connection lookup logic out of DID exchange service to common package, so that it can be reused by client packages to lookup connection records.
- closes #1015

TODO: 

- save/updates logic to be moved as part of https://github.com/hyperledger/aries-framework-go/issues/1021
- connection and DID store to be merged as part of https://github.com/hyperledger/aries-framework-go/issues/1021

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

File: pkg/didcomm/common/service/connection_record.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+// NewConnectionRecorder returns new connection recorder instance
+func NewConnectionRecorder(p provider) (*ConnectionRecorder, error) {
+	store, err := p.StorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	transientStore, err := p.TransientStorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
```
Comment (soluchok) [inline pkg/didcomm/common/service/connection_record.go]: same error messages?

File: pkg/didcomm/common/service/connection_record.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
+		}
+
+		return nil, err
+	}
+
+	return rec, nil
+}
```
Comment (soluchok) [inline pkg/didcomm/common/service/connection_record.go]: looks better
```
func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
   rec, err := getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.store)
    if errors.Is(err, storage.ErrDataNotFound) {
		return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
	}

	return rec, err
}
```

File: pkg/didcomm/protocol/didexchange/persistence.go [hunk 3]
```diff
@@ -12,56 +12,33 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+
```
Comment (soluchok) [inline pkg/didcomm/protocol/didexchange/persistence.go]: extra line

File: pkg/didcomm/common/service/connection_record.go (hunk 2 above)
Comment (sudeshrshetty) [reply pkg/didcomm/common/service/connection_record.go]: will make sure it doesn't panic when err = nil

File: pkg/didcomm/common/service/connection_record.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		if _, ok := keys[string(transientItr.Key())]; !ok {
+			var record ConnectionRecord
+
+			err := json.Unmarshal(transientItr.Value(), &record)
+			if err != nil {
+				return nil, fmt.Errorf("query connection records from transient store : %w", err)
+			}
+
+			records = append(records, &record)
+		}
```
Comment (soluchok) [inline pkg/didcomm/common/service/connection_record.go]: I know it is not your code, but better to avoid nested if statements e.g looks better to me

```
if _, ok := keys[string(transientItr.Key())]; ok {
    continue
}

var record ConnectionRecord

if err := json.Unmarshal(transientItr.Value(), &record); err != nil {
    return nil, fmt.Errorf("query connection records from transient store : %w", err)
}

records = append(records, &record)
```

Comment (sudeshrshetty) [reply pkg/didcomm/common/service/connection_record.go]: Just moved existing code, will fix this.

Comment (sudeshrshetty) [reply pkg/didcomm/common/service/connection_record.go]: done

File: pkg/didcomm/common/service/connection_record.go (hunk 2 above)
Comment (sudeshrshetty) [reply pkg/didcomm/common/service/connection_record.go]: done

File: pkg/didcomm/common/service/connection_record.go (hunk 1 above)
Comment (sudeshrshetty) [reply pkg/didcomm/common/service/connection_record.go]: done

File: pkg/didcomm/protocol/didexchange/persistence.go (hunk 3 above)
Comment (sudeshrshetty) [reply pkg/didcomm/protocol/didexchange/persistence.go]: done

File: pkg/client/didexchange/client.go [hunk 5]
```diff
@@ -67,6 +68,9 @@ type protocolService interface {
 	// CreateImplicitInvitation creates implicit invitation. Inviter DID is required, invitee DID is optional.
 	// If invitee DID is not provided new peer DID will be created for implicit invitation exchange request.
 	CreateImplicitInvitation(inviterLabel, inviterDID, inviteeLabel, inviteeDID string) (string, error)
+
+	// SaveInvitation saves given invitation in did-exchange service connection store
+	SaveInvitation(invitation *didexchange.Invitation) error
```
Comment (troyronda) [inline pkg/client/didexchange/client.go]: Why is this needed in client?

Perhaps it move again with the next task.

File: pkg/client/didexchange/models.go [hunk 6]
```diff
@@ -6,7 +6,11 @@ SPDX-License-Identifier: Apache-2.0
 
 package didexchange
 
-import "github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/didexchange"
+import (
+	"github.com/hyperledger/aries-framework-go/pkg/common/connectionstore"
+
```
Comment (troyronda) [inline pkg/client/didexchange/models.go]: Extra line

File: pkg/common/connectionstore/connection_record.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+	connRecordBytes, err := c.transientStore.Get(k)
+	if err != nil {
+		return nil, fmt.Errorf("get connection record by connectionID: %w", err)
+	}
+
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// Store returns handle to underlying permanent store
```
Comment (troyronda) [inline pkg/common/connectionstore/connection_record.go]: Are these used?

Comment (sudeshrshetty) [reply pkg/common/connectionstore/connection_record.go]: Yes, private `connectionStore` inside didexchange service  which wraps this struct calls Put() & Get() on these store instances ( Store() & TransientStore() functions).
These two `Store() and TransientStore()` will be removed once we move those custom save/update logic out of didexchange service package.

File: pkg/client/didexchange/models.go (hunk 6 above)
Comment (sudeshrshetty) [reply pkg/client/didexchange/models.go]: done

File: pkg/client/didexchange/client.go (hunk 5 above)
Comment (sudeshrshetty) [reply pkg/client/didexchange/client.go]: 'CreateInvitation' logic is actually inside didexchange client.
So DIDExchangeService exports SaveInvitation() function to retain `save` logic inside service.
Again, this logic will be removed once writable '*connectionStore' is moved out of didexchange.
